### PR TITLE
fix(common): titlecase pipe incorrectly handling numbers

### LINKED ISSUE #43459: TitleCasePipe performs incorrect transformation for word starting with digit
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

Word such as `2nd` is transformed to `2Nd` which is incorrect. 

Expected Behavior is that the word remains unchanged. If input word was `2ND` it should change to lowercase `2nd`

### Please provide a link to a minimal reproduction of the bug

_No response_

### Please provide the exception or error you saw

```true
No exception
```

### Please provide the environment you discovered this bug in

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by crisbeto**
Fixes that the `titlecase` pipe wasn't picking up numbers which meant that cases like `1st` would be transformed to `1St`.

Fixes #43459.

### CODE REVIEW

File: packages/common/test/pipes/case_conversion_pipes_spec.ts [hunk 1]
```diff
@@ -90,16 +90,23 @@ import {LowerCasePipe, TitleCasePipe, UpperCasePipe} from '@angular/common';
       expect(pipe.transform('éric')).toEqual('Éric');
     });
 
+    it('should handle numbers at the beginning of words', () => {
+      expect(pipe.transform('frodo was 1st and bilbo was 2nd'))
+          .toEqual('Frodo Was 1st And Bilbo Was 2nd');
```
Comment (petebacondarwin) [inline packages/common/test/pipes/case_conversion_pipes_spec.ts:95]: Could you also test lowering the case of the second character after a number?
E.g. `transform('1ST')` -> `1st`

Comment (crisbeto) [reply packages/common/test/pipes/case_conversion_pipes_spec.ts:95]: Done.

**Review by petebacondarwin — APPROVED** [review]
LGTM!

**Review by AndrewKushnir — COMMENTED** [review]
@crisbeto would this change be compatible with the ES2018 version of this RegEx (i.e. `/\p{L}/u`)?

**Comment by AndrewKushnir** [comment]
I've looked at the changes and the comment above this RegExp and I think we should not change it manually, so we can replace it with the `/\p{L}/u` version once we transition to ES2018. Instead we can try to come up with a RegExp that would leverage ES2018 capabilities (+ extra processing for a specific case that we handle) and transpile that to ES2015 version. @crisbeto what do you think?

**Comment by crisbeto** [comment]
That would be the equivalent to something like `/\p{L}\p{N}/u`, but that makes the regex even more massive. See https://mothereff.in/regexpu#input=var+regex+%3D+%2F%5Cp%7BL%7D%5Cp%7BN%7D%2Fu%3B&unicodePropertyEscape=1.

It adds another 4kb to the regex so I'm not sure that the trade-off is worth it.

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

File: packages/common/src/pipes/case_conversion_pipes.ts [hunk 2]
```diff
@@ -51,7 +51,7 @@ export class LowerCasePipe implements PipeTransform {
...
...E38\uDE40-\uDE5E\uDED0-\uDEED\uDF00-\uDF2F\uDF40-\uDF43\uDF63-\uDF77\uDF7D-\uDF8F]|\uD81B[\uDF00-\uDF44\uDF50\uDF93-\uDF9F\uDFE0\uDFE1]|\uD821[\uDC00-\uDFEC]|\uD822[\uDC00-\uDEF2]|\uD82C[\uDC00-\uDD1E\uDD70-\uDEFB]|\uD82F[\uDC00-\uDC6A\uDC70-\uDC7C\uDC80-\uDC88\uDC90-\uDC99]|\uD835[\uDC00-\uDC54\uDC56-\uDC9C\uDC9E\uDC9F\uDCA2\uDCA5\uDCA6\uDCA9-\uDCAC\uDCAE-\uDCB9\uDCBB\uDCBD-\uDCC3\uDCC5-\uDD05\uDD07-\uDD0A\uDD0D-\uDD14\uDD16-\uDD1C\uDD1E-\uDD39\uDD3B-\uDD3E\uDD40-\uDD44\uDD46\uDD4A-\uDD50\uDD52-\uDEA5\uDEA8-\uDEC0\uDEC2-\uDEDA\uDEDC-\uDEFA\uDEFC-\uDF14\uDF16-\uDF34\uDF36-\uDF4E\uDF50-\uDF6E\uDF70-\uDF88\uDF8A-\uDFA8\uDFAA-\uDFC2\uDFC4-\uDFCB]|\uD83A[\uDC00-\uDCC4\uDD00-\uDD43]|\uD83B[\uDE00-\uDE03\uDE05-\uDE1F\uDE21\uDE22\uDE24\uDE27\uDE29-\uDE32\uDE34-\uDE37\uDE39\uDE3B\uDE42\uDE47\uDE49\uDE4B\uDE4D-\uDE4F\uDE51\uDE52\uDE54\uDE57\uDE59\uDE5B\uDE5D\uDE5F\uDE61\uDE62\uDE64\uDE67-\uDE6A\uDE6C-\uDE72\uDE74-\uDE77\uDE79-\uDE7C\uDE7E\uDE80-\uDE89\uDE8B-\uDE9B\uDEA1-\uDEA3\uDEA5-\uDEA9\uDEAB-\uDEBB]|\uD869[\uDC00-\uDED6\uDF00-\uDFFF]|\uD86D[\uDC00-\uDF34\uDF40-\uDFFF]|\uD86E[\uDC00-\uDC1D\uDC20-\uDFFF]|\uD873[\uDC00-\uDEA1\uDEB0-\uDFFF]|\uD87A[\uDC00-\uDFE0]|\uD87E[\uDC00-\uDE1D])\S*/g;
```
Comment (gkalpak) [inline packages/common/src/pipes/case_conversion_pipes.ts:54]: WRT https://github.com/angular/angular/pull/43476#issuecomment-921330461:

I don't think we need `\p{N}`. Using `0-9` would be fine (in ES2018). So, we can keep the regex as is now and just update the comment to mention it is the equivalent of `/[0-9\p{L}]\S*/gu` (so we can update it to the correct version in the future).

BTW, the comment currently implies that the regex can be replaced with `/\p{L}\S*/gu` for ES2018, which is no longer true after this PR. So, we need to update the comment (one way or another).

**Comment by crisbeto** [comment]
It's worth noting that the previous regex wasn't pasted verbatim from that link. The site transpiles `\S*` to `(?:[\0-\x08\x0E-\x1F!-\x9F\xA1-\u167F\u1681-\u1FFF\u200B-\u2027\u202A-\u202E\u2030-\u205E\u2060-\u2FFF\u3001-\uD7FF\uE000-\uFEFE\uFF00-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])*` whereas our regex had `\S*` stuck to the end of it. I'll keep the same pattern for now since it's smaller.

**Review by AndrewKushnir — APPROVED** [review]
Thanks for updating docs 👍 I'm also ok to update the RegExp, but keep the docs up-to-date and make it possible to replace it with the short version later. 

> It's worth noting that the previous regex wasn't pasted verbatim from that link. The site transpiles \S* to (?:[\0-\x08\x0E-\x1F!-\x9F\xA1-\u167F\u1681-\u1FFF\u200B-\u2027\u202A-\u202E\u2030-\u205E\u2060-\u2FFF\u3001-\uD7FF\uE000-\uFEFE\uFF00-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])* whereas our regex had \S* stuck to the end of it. I'll keep the same pattern for now since it's smaller.

Could you please add this note to the comment above the RegExp as well (so we have extra context when we decide to replace it with the short version)?

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:397651840:BASE:397651833:1632098862035:2a7d283).

**Comment by AndrewKushnir** [comment]
FYI regular presubmits went well, starting a [global one](https://test.corp.google.com/OCL:397651840:BASE:397665929:1632108986957:d090b4af) (internal-only link) to see if there are any targets that might require cleanup.

**Comment by AndrewKushnir** [comment]
FYI presubmits (including a global one) are successful, removing the "presubmit" label.

**Pull request closed without merging**